man/consoleLogger.3
.\" Written  Oct 2012 by Olivier Charloton (oliviers.open.source.softwares@gmail.com)
.\" heavily based on the work of Steve Greenland (stevegr@neosoft.com) on syslog man.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.TH CONSOLELOGGER 3 2012-10-06 "dbgflags" "Dbgflags Programmer's Manual"
.SH NAME
openConsoleLogger, setConsoleLoggerOpt, consoleLogger, vconsoleLogger \- log messages to a console.
.SH SYNOPSIS
.B #include <dbgflags/loggers.h>
.sp
.BI "void openConsoleLogger(const char *" ident ", int " option ", int " facility ");"
.br
.BI "void consoleLogger(int " priority ", const char *" format ", ...);"
.br
.BI "void setConsoleLoggerOpt(const char *" ident ", int " option ", int " facility ");"
.sp
.B #include <stdarg.h>
.sp
.BI "void vconsoleLogger(int " priority ", const char *" format ", va_list " ap ");"
.br
.SH DESCRIPTION
.BR openConsoleLogger ()
initialize the console logger's parameters .  The string pointed
to by
.I ident
is prepended to every message, and is typically set to the program name.
The
.I option
argument specifies flags which control the operation of
.BR openConsoleLogger ()
and subsequent calls to
.BR consoleLogger ().
The
.I facility
argument establishes a default to be used if
none is specified in subsequent calls to
.BR consoleLogger ().
Values for
.I option
and
.I facility
are given below.  The use of 
.BR openConsoleLogger ()
is optional;
.sp
.BR consoleLogger ()
generates a log message, which will be written to the program's console (if any).
The
.I priority
argument is formed by ORing the
.I facility
and the 
.I level
values (explained below).  The remaining arguments
are a 
.IR format ,
as in 
.BR printf (3)
and any arguments required by the 
.IR format , 
except that the two character sequence %m will be replaced by
the error message string 
.IR strerror ( errno ).
A trailing newline is added when needed.

The function
.BR vconsoleLogger ()
performs the same task as
.BR consoleLogger ()
with the difference that it takes a set of arguments which have
been obtained using the
.BR stdarg (3)
variable argument list macros.

.BR setConsoleLoggerOpt()
Deprecated version of the openConsoleLogger function, present for API compatibility reason. 

.SH PARAMETERS
This section lists the parameters used to set the values of 
.IR option , " facility" ", and " priority 
(sames as 
.BR syslog(3)
).
.SS option
The
.I option
argument to 
.BR openConsoleLogger ()
is an OR of any of these:
.TP
.B LOG_CONS
Ignored.
.TP
.B LOG_NDELAY
Ignored.
.TP
.B LOG_NOWAIT
Ignored.
.TP
.B LOG_ODELAY
The converse of
.BR LOG_NDELAY ;
ignored too.
.TP
.B LOG_PERROR
Ignored.
.TP
.B LOG_PID
Include PID with each message.
.TP
.B LOG_TID
(dbgflags's extension) Include Thread's identity (TID) with each message.
.TP
.B LOG_RDTSC
(dbgflags's extension) Log the number of clock cycles, since the CPU was powered up or reset, with each message. 
Warning: this value is different between every core and CPU, when running on more then one, use the LOG_CLOCK instead.
.TP 
.B LOG_LEVEL
(dbgflags's extension) Log the message level as a string with each message.
.TP
.B LOG_CLOCK
(dbgflags's extension) Log a time, that represents monotonic time since some unspecified starting point, with each message.

.SS facility 
The 
.I facility
argument is used to specify what type of program is logging the message.
This lets the configuration file specify that messages from different
facilities will be handled differently.
.TP
.B LOG_AUTH
security/authorization messages (DEPRECATED Use 
.B LOG_AUTHPRIV 
instead)
.TP
.B LOG_AUTHPRIV
security/authorization messages (private) 
.TP
.B LOG_CRON
clock daemon 
.RB ( cron " and " at )
.TP
.B LOG_DAEMON
system daemons without separate facility value
.TP
.B LOG_FTP
ftp daemon
.TP
.B LOG_KERN
kernel messages
.TP
.BR LOG_LOCAL0 " through " LOG_LOCAL7
reserved for local use
.TP
.B LOG_LPR
line printer subsystem
.TP
.B LOG_MAIL
mail subsystem
.TP
.B LOG_NEWS
USENET news subsystem
.TP
.B LOG_SYSLOG
messages generated internally by
.B syslogd 
.TP
.BR LOG_USER " (default)"
generic user-level messages 
.TP
.B LOG_UUCP
UUCP subsystem

.SS level
This determines the importance of the message.  The levels are, in order
of decreasing importance:
.TP
.B LOG_EMERG
system is unusable
.TP
.B LOG_ALERT
action must be taken immediately
.TP
.B LOG_CRIT
critical conditions
.TP
.B LOG_ERR
error conditions
.TP
.B LOG_WARNING
warning conditions
.TP
.B LOG_NOTICE
normal, but significant, condition
.TP
.B LOG_INFO
informational message
.TP
.B LOG_DEBUG
debug-level message
.LP
The functions
.BR setlogmaskex (3)
or
.BR setlogmask (3)
can be used to restrict logging to specified levels only.
.SH NOTES
The parameter
.I ident
in the call of
.BR openConsoleLogger ()
is stored as-is. Thus, if the string it points to
is changed,
.BR consoleLogger ()
may start prepending the changed string, and if the string
it points to ceases to exist, the results are undefined.
Use a string constant.
.LP
Never pass a string with user-supplied data as a format, use the following instead:
.RS
consoleLogger(priority, "%s", string);

.SH "SEE ALSO"
.BR logger (1),
.BR syslog (3),
.BR syslogex (3),
.BR setlogmask (3),



NEWS
- A few corrections to build on a Debian 6 64bits.
- pkgconfig test program build corrected when the library hase not been already installed.
- the gettid internal function use the syscall function (instead of the "int 0x80") when building on a 2.6.x Linux kernel.
- Buffer overflow correction in consoleLogger & fileLogger (use a stream buffer now instead of a buffer on the stack).
  and new test programs to check the library's logger with message of size up to 4096 bytes.
- thread cancellation points management in fileLogger added (useless in the consoleLogger).
- added instructions to free the used buffer in the thread cancellation points management of the syslogex logger (glibc bug ?)
- logger sending trace to one file per thtread

Remarques:
- Files'repository moved to GIT and "public" version published to github
- using redmine to manage evolutions (ER) and anomalies reports (AR).


libdbgflags/consoleLogger.c
#include "config.h"
#include <stdio.h>
#include <stdarg.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>
#include <dbgflags/syslogex.h>
#include <dbgflags/loggers.h>
#include <dbgflags/debug_macros.h>
#include "system.h"


static int	LogStat = 0;		/* status bits, set by openlog() */
static const char *LogTag = NULL;		/* string to tag the entry with */
static int	LogFacility = LOG_USER;	/* default facility code */
//static int	LogMask = 0xff;		/* mask of priorities to be logged */

void openConsoleLogger(const char *ident, int logstat, int logfac) {
    if (ident != NULL) {
		LogTag = ident;
    }
    LogStat = logstat;
    if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0) {
		LogFacility = logfac;
    }
}

extern void setConsoleLoggerOpt(const char *ident, int logstat, int logfac) __attribute((alias("openConsoleLogger")));
   
void vconsoleLogger(int priority, const char *format, va_list optional_arguments) {
    /*int n = 0;*/
    const int saved_errno = errno;
    const int LogMask = setlogmask(0);

    /* no cancellation point is currently used in this function 
    * (according to Advanced Programming in the Unix Environment 2nd ed p411) 
     * so there is no thread cancellation clean-up handlers defined      
     */
    
    /* Check for invalid bits. */
    if (unlikely(priority & ~(LOG_PRIMASK | LOG_FACMASK))) {
        /*syslog(INTERNALLOG,
               "syslog: unknown facility/priority: %x", pri);*/
        WARNING_MSG("unknown facility/priority: %x", priority);
        priority &= LOG_PRIMASK | LOG_FACMASK;
    }

    /* Check priority against setlogmask values. */  
    if ((LOG_MASK (LOG_PRI (priority)) & LogMask) != 0) {
        char *buf = 0;
        size_t bufsize = 1024;
        /*char logFormat[1024];
        char *cursor = logFormat;*/
        FILE *f = open_memstream(&buf, &bufsize);
	if (f != NULL) {                
	  struct tm now_tm;
	  time_t now;

          (void) time(&now);
          /*cursor += strftime(cursor,sizeof(logFormat),"%h %e %T ",localtime_r(&now, &now_tm));*/
	  f->_IO_write_ptr += strftime(f->_IO_write_ptr,f->_IO_write_end - f->_IO_write_ptr,"%h %e %T ",localtime_r(&now, &now_tm));

	  if (LogTag) {
	    //cursor += sprintf (cursor,"%s: ",LogTag);
	    fprintf(f,"%s: ",LogTag);
	  }

	  if (LogStat & LOG_PID) {
	      if (LogStat & LOG_TID) {
		  const pid_t tid = gettid();
		  /*cursor += sprintf (cursor, "[%d:%d]", (int) getpid (),(int) tid);*/
		  fprintf(f,"[%d:%d]", (int) getpid (),(int) tid);
	      } else {
		  /*cursor += sprintf (cursor, "[%d]", (int) getpid ());*/
		  fprintf(f,"[%d]", (int) getpid ());
	      }             
	  }

	  if (LogStat & LOG_RDTSC) {
	      const unsigned long long int  t = rdtsc();
	      /*cursor += sprintf (cursor, "(%llu)",t);*/
	      fprintf(f,"(%llu)",t);
	  } /* (LogStat & LOG_RDTSC) */

	  if (LogStat & LOG_CLOCK) {
	      #if HAVE_CLOCK_GETTIME
		  struct timespec timeStamp;
		  if (clock_gettime(CLOCK_MONOTONIC,&timeStamp) == 0) {
		      /*cursor += sprintf (cursor, "(%lu.%.9d)",timeStamp.tv_sec,timeStamp.tv_nsec);*/
		      fprintf(f,"(%lu.%.9d)",timeStamp.tv_sec,timeStamp.tv_nsec);
		  } else {
		      const int error = errno;
		      ERROR_MSG("clock_gettime CLOCK_MONOTONIC error %d (%m)",error);
		  }
	      #else
		  static unsigned int alreadyPrinted = 0; /* to avoid to print this error msg on each call */
		  if (unlikely(0 == alreadyPrinted)) {
		      ERROR_MSG("clock_gettime  not available on this system");
		      alreadyPrinted = 1;
		  }
	      #endif
	  } /* (LogStat & LOG_CLOCK) */

	  if (LogStat & LOG_LEVEL) {
		switch(LOG_PRI(priority)) {
		case LOG_EMERG:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Emergency * %s",format);*/
			fprintf(f,"[EMERG] %s",format);
			break;
		case LOG_ALERT:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Alert * %s",format);*/
			fprintf(f,"[ALERT] %s",format);
			break;
		case LOG_CRIT:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Critical * %s",format);*/
			fprintf(f,"[CRIT] %s",format);
			break;
		case LOG_ERR:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Error * %s",format);*/
			fprintf(f,"[ERROR] %s",format);
			break;
		case LOG_WARNING:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Warning * %s",format);*/
			fprintf(f,"[WARNING] %s",format);
			break;
		case LOG_NOTICE:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Notice * %s",format); */
			fprintf(f,"[NOTICE] %s",format);
			break;
		case LOG_INFO:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Info * %s",format);*/
			fprintf(f,"[INFO] %s",format);
			break;
		case LOG_DEBUG:
			/*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"* Debug * %s",format);*/
			fprintf(f,"[DEBUG] %s",format);
			break;
		default:
			/*cursor += sprintf(cursor,"* <%d> * %s",priority,format);*/
			fprintf(f,"[<%d>] %s",priority,format);
		} /* switch(priority) */        
	  } else { /* (LogStat & LOG_LEVEL) */
		  /*cursor += snprintf(cursor,sizeof(logFormat) - (cursor - logFormat),"%s",format);*/
		  fprintf(f,"%s",format);
	  }

	  /*n =*/ /*vfprintf(stderr,logFormat,optional_arguments);*/
	  /* Close the memory stream; this will finalize the data
           into a malloc'd buffer in BUF.  */
	  fclose(f);
	  errno = saved_errno; /* restore errno for %m format */
	  vfprintf(stderr,buf,optional_arguments);
	  free(buf);
	} else {
	  /* We cannot get a stream: try to write directly to the console (warning may be splitted by other msg) */
	  struct tm now_tm;
	  time_t now;
	  char buffer[20];	  

	  WARNING_MSG("failed to get stream buffer, using the console without buffering instead");
          (void) time(&now);
	  strftime(buffer,sizeof(buffer),"%h %e %T ",localtime_r(&now, &now_tm));
	  fprintf(stderr,"%s",buffer);
	  
	  if (LogTag) {	    
	    printf("%s: ",LogTag);
	  }

	  if (LogStat & LOG_PID) {
	      if (LogStat & LOG_TID) {
		  const pid_t tid = gettid();		  
		  fprintf(stderr,"[%d:%d]", (int) getpid (),(int) tid);
	      } else {		  
		  fprintf(stderr,"[%d]", (int) getpid ());
	      }             
	  }

	  if (LogStat & LOG_RDTSC) {
	      const unsigned long long int  t = rdtsc();	      
	      fprintf(stderr,"(%llu)",t);
	  } /* (LogStat & LOG_RDTSC) */
	  
	  if (LogStat & LOG_CLOCK) {
	      #if HAVE_CLOCK_GETTIME
		  struct timespec timeStamp;
		  if (clock_gettime(CLOCK_MONOTONIC,&timeStamp) == 0) {		      
		      fprintf(stderr,"(%lu.%.9d)",timeStamp.tv_sec,timeStamp.tv_nsec);
		  } else {
		      const int error = errno;
		      ERROR_MSG("clock_gettime CLOCK_MONOTONIC error %d (%m)",error);
		  }
	      #else
		  static unsigned int alreadyPrinted = 0; /* to avoid to print this error msg on each call */
		  if (unlikely(0 == alreadyPrinted)) {
		      ERROR_MSG("clock_gettime  not available on this system");
		      alreadyPrinted = 1;
		  }
	      #endif
	  } /* (LogStat & LOG_CLOCK) */
	  
	  if (LogStat & LOG_LEVEL) {
		switch(LOG_PRI(priority)) {
		case LOG_EMERG:			
			fprintf(stderr,"* Emergency * ");
			break;
		case LOG_ALERT:			
			fprintf(stderr,"* Alert * ");
			break;
		case LOG_CRIT:
			fprintf(stderr,"* Critical * ");
			break;
		case LOG_ERR:			
			fprintf(stderr,"* Error * ");
			break;
		case LOG_WARNING:			
			fprintf(stderr,"* Warning * ");
			break;
		case LOG_NOTICE:			
			fprintf(stderr,"* Notice * ");
			break;
		case LOG_INFO:			
			fprintf(stderr,"* Info * ");
			break;
		case LOG_DEBUG:			
			fprintf(stderr,"* Debug * ");
			break;
		default:			
			fprintf(f,"* <%d> * ",priority);
		} /* switch(priority) */        
	  } 
	  
	  vfprintf(stderr,format,optional_arguments);
	}
    } /* ((LOG_MASK (LOG_PRI (priority)) & LogMask) != 0) */
    /* message has not to be displayed because of the current LogMask and its priority */

    /*return n;*/
}

void consoleLogger(int priority, const char *format, ...) {    
    va_list optional_arguments;
    va_start(optional_arguments, format);
    vconsoleLogger(priority, format,  optional_arguments);
    va_end(optional_arguments);
    /*return n;*/
}

#include <dbgflags/ModuleVersionInfo.h>
MODULE_NAME(consoleLogger);
MODULE_AUTHOR(Olivier Charloton);
MODULE_VERSION(0.1);
MODULE_FILE_VERSION(1.0);
MODULE_DESCRIPTION(logger to error console with the same signature than syslog);
MODULE_COPYRIGHT(LGPL);


